### PR TITLE
Added MediaScanner for scanning images

### LINKED ISSUE #1739: Pictures are not saved in an easy accessible place
If I "Save a media" (a picture here), it seems to put it in a place which is not easily available. I expect to have it in "Downloads" or "Pictures" like if I would take a picture with the camera or download a file with the browser. I have an Android 10.

### PR DESCRIPTION
**Pull request by abhinav12k**
**Explanation**
Fixes #1739 

Added a MediaScanner to scan images and display them in the gallary.

**Current Output** 

https://user-images.githubusercontent.com/51455561/109339429-a80bce80-788d-11eb-8985-aed98d8050b5.mp4

Tested on Asus Zenfone Max Pro M1(Android 11)

### CODE REVIEW

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixWebView.kt [hunk 1]
```diff
@@ -180,6 +180,15 @@ open class KiwixWebView @SuppressLint("SetJavaScriptEnabled") constructor(
           Log.w("kiwix", "Couldn't save image", e)
           instance.toast(R.string.save_media_error)
         }
+
+        MediaScannerConnection.scanFile(
+          instance.applicationContext,
+          arrayOf(root.toString()),
+          null
+        ) { path, uri ->
+          // Log.i("ExternalStorage", "Scanned $path:")
```
Comment (mhutti1) [inline core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixWebView.kt:189]: Can we remove these comments

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixWebView.kt [hunk 2]
```diff
@@ -180,6 +180,15 @@ open class KiwixWebView @SuppressLint("SetJavaScriptEnabled") constructor(
           Log.w("kiwix", "Couldn't save image", e)
           instance.toast(R.string.save_media_error)
         }
+
+        MediaScannerConnection.scanFile(
```
Comment (mhutti1) [inline core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixWebView.kt:184]: Add a comment explaining what is happening here.

**Review by mhutti1 — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixWebView.kt (hunk 1 above)
Comment (abhinav12k) [reply core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixWebView.kt:189]: Ya sure, I will remove them

File: core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixWebView.kt (hunk 2 above)
Comment (abhinav12k) [reply core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixWebView.kt:185]: I have added a new commit can you please take a look.

**Comment by s-ayush2903** [comment]
Can you please add the recording of the functionality implemented(as adding visuals is are a part of PR template and is highly encouraged, but can be avoided in some non-related tickets) to make sure it works as expected, and also to speed up review process.

**Comment by s-ayush2903** [comment]
Can you please add the recording of the functionality implemented(as adding visuals is a part of PR template and is highly encouraged, but can be avoided in some non-related tickets) to make sure it works as expected, and also to speed up review process.

**Comment by abhinav12k** [comment]
Hi @s-ayush2903, I have added a recording of the current behavior please check it out.

**Review by s-ayush2903 — DISMISSED** [review]
Looks good to me!

**Comment by kelson42** [comment]
@s-ayush2903 so you really confirm that this PR fixes the bug reported? This is not absilutky obvious to me so far.

**Comment by s-ayush2903** [comment]
@kelson42 Now that I look at the issue title & description again, I believe that this PR fixes the issue to an extent, it now exposes the media via making it visible to other applications for open to discovery, previously, images weren't being displayed in the gallery of my device(which isn't the case now). The only work left is to provide the directory a better name instead `org.kiwix.mobile`, I think `Kiwix` should be fine. And to save images in Pictures/Downloads dir we'll need write access to 'em(which isn't needed as we can provide users with the folder of the name we prefer in their gallery, and this'll keep our content separate from other media stuff that users download on the go) as Frans mentioned.

**Comment by gouri-panda** [comment]
@mhutti1 @abdulwd Is it better if we download photos by android's `DownloadService`. So that we don't have to configure the path and other settings it will be done by android itself. By default, downloads are saved to a generated filename in the android shared download folder. The downloaded file is not scanned by `MediaScanner`. But it can be made scannable by calling `allowScanningByMediaScanner()`.

**Comment by gouri-panda** [comment]
Also @abhinav12k Please don't resolve conversations by yourself. It generally is resolved by those who open the conversation. So that it'll help to review them. I'm going to unresolved the conversation ;)

**Comment by kelson42** [comment]
> The only work left is to provide the directory a better name instead `org.kiwix.mobile`, I think `Kiwix` should be fine.

Thank you for your explanation. I support the change of directory name.

**Comment by gouri-panda** [comment]
>@mhutti1 @abdulwd Is it better if we download photos by android's DownloadService. So that we don't have to configure the path and other settings it will be done by android itself. By default, downloads are saved to a generated filename in the android shared download folder. The downloaded file is not scanned by MediaScanner. But it can be made scannable by calling allowScanningByMediaScanner().
I forgot to mention that using this API. We must have `WRITE_EXTERNAL_STORAGE` permission.

**Comment by abhinav12k** [comment]
> Also @abhinav12k Please don't resolve conversations by yourself. It generally is resolved by those who open the conversation. So that it'll help to review them. I'm going to unresolved the conversations ;)

Sorry, I didn't know about this I thought that as i corrected that in the latest commit i can resolve conversation there 😅.

**Comment by abhinav12k** [comment]
> > The only work left is to provide the directory a better name instead `org.kiwix.mobile`, I think `Kiwix` should be fine.
> 
> Thank you for your explanation. I support the change of directory name.

Should i try to change the directory name?

**Comment by kelson42** [comment]
@abhinav12k yes please

**Comment by kelson42** [comment]
What is the status here, it sounds it needs a new review?

**Comment by abhinav12k** [comment]
I tried a few approaches but I am unable to reach the desired result. I have mentioned it over slack too. 
Approaches that I tried ->
1. Changed the directory name to kiwix but after changing the directory name the media scanner is not able to detect the images from the new directory. 
2. As the media scanner was not able to detect pictures in the directory. I tried to use the MediaStore api but that also didn't work. I tried to take help from [this article](https://medium.cobeisfresh.com/taming-file-storage-on-android-part-2-4afe9dfd9b6e) and also from [stackoverflow](https://stackoverflow.com/questions/56904485/how-to-save-an-image-in-android-q-using-mediastore).

**Comment by kelson42** [comment]
@gouri-panda What should we do now?

**Comment by gouri-panda** [comment]
@abhinav12k Which version of android you are using?

**Comment by gouri-panda** [comment]
also, Acc. to documentation `instance.getExternalFilesDir(Environment.DIRECTORY_PICTURES), "Kiwix")` If the user uninstalls the app these files (in this case images) will be deleted. I think we don't want to do that :)

**Comment by gouri-panda** [comment]
@abhinav12k Feel free to open this PR when you're ready.

**Pull request closed without merging**